Add feature limit support for feature requests

nyalldawson opened the pull request:
Here's an implementation of adding limits to QgsFeatureRequest. Setting a limit will restrict the maximum number of features returned by the request. This limit can be passed on to the provider for limiting the query results on the server side, yielding performance benefits when only a set number of features is required.

I'm looking for feedback on the approach and whether this is the right way to implement this.

nyalldawson committed:
[FEATURE] Feature limit support for feature requests

Limits the maximum number of features returned by the iterator.
Some providers (postgres, spatialite, MS SQL) pass the limit on
to the provider to result in faster queries.

elpaso commented:
Could this help with http://hub.qgis.org/issues/12926 ? Maybe we could fetch a limited number of features and fetch another batch when the user scrolls down.

michaelkirk commented on tests/src/python/providertestbase.py:
```diff
@@ -127,6 +127,22 @@ def testRectAndExpression(self):
+    def testGetFeaturesLimit(self):
+        it = self.provider.getFeatures(QgsFeatureRequest().setFeatureLimit(2))
+        features = [f['pk'] for f in it]
+        assert len(features) == 2, 'Expected two features, got {} instead'.format(features)
```
Do you want to print `len(features)` in the error message?

Also, consider using `self.assertEqual(len(features), 2)` which gives a pretty decent, though generic error message, out of the box. The benefit being that you don't have to worry about composing a (maybe buggy) error message.

nyalldawson commented:
Updated version, considering earlier comments. I've also fixed the todo- regarding disabling the limit when the server can't handle a request completely.

nyalldawson commented:
Any objections if I merge?

nyalldawson commented:
Merged in edb16d0014dcc0f9017c0711554dfb7aa7a51bb9
